process/block/shardInfo.go

```diff
@@ -113,24 +115,28 @@ func (sic *ShardInfoCreateData) CreateShardInfoFromLegacyMeta(
 	return shardInfo, nil
 }
 
-func (sic *ShardInfoCreateData) createShardInfoFromHeader(shardHeader data.HeaderHandler, hdrHash []byte) ([]data.ShardDataHandler, error) {
+func (sic *ShardInfoCreateData) createShardInfoFromHeader(
+	shardHeader data.HeaderHandler,
+	hdrHash []byte,
+) (data.ShardDataProposalHandler, []data.ShardDataHandler, error) {
 	if check.IfNil(shardHeader) {
-		return nil, process.ErrNilHeaderHandler
+		return nil, nil, process.ErrNilHeaderHandler
 	}
 	if len(hdrHash) == 0 {
-		return nil, process.ErrInvalidHash
+		return nil, nil, process.ErrInvalidHash
 	}
 
 	hasMissingShardHdrProof := shardHeader.GetNonce() >= 1 && !sic.proofsPool.HasProof(shardHeader.GetShardID(), hdrHash)
 	if hasMissingShardHdrProof {
-		return nil, fmt.Errorf("%w for shard header with hash %s", process.ErrMissingHeaderProof, hex.EncodeToString(hdrHash))
+		return nil, nil, fmt.Errorf("%w for shard header with hash %s", process.ErrMissingHeaderProof, hex.EncodeToString(hdrHash))
 	}
 
 	if !shardHeader.IsHeaderV3() {
-		return sic.createShardDataFromLegacyHeader(shardHeader, hdrHash)
+		shardData, err := sic.createShardDataFromLegacyHeader(shardHeader, hdrHash)
+		return &block.ShardDataProposal{}, shardData, err // no shard data proposal for legacy header
```

Comment:
Since createShardInfoFromHeader is called to create the shardInfo on a MetaV3 header, I think we should probably track the shard headersV2 also as proposals (they are proposed and executed at the same time). Then we won't miss any headers when transitioning to the V3 meta header.

Reply:
updated